solutions/authorization/index.asciidoc

```diff
@@ -0,0 +1,83 @@
+
+.Security terms related to authorization
+[options="header", cols="15%,85%"]
+|=======================
+|*Term*|*Meaning and comment*
+|Permission|A permission is an object that allows a principal to perform an operation in the system. This permission can be _granted_ (give) or _revoked_ (taken away). Sometimes people also use the term _right_ what is actually wrong as a right (such as the right to be free) can not be revoked.
+|Group|We use the term group in this context for an object that contains permissions. A group may also contain other groups. Then the group represents the set of all recursively contained permissions.
+|Role|We consider a role as a specific form of group that also contains permissions. A role identifies a specific function of a principal. A user can act in a role.
+
+For simple scenarios a principal has a single role associated. In more complex situations a principal can have multiple roles but has only one active role at a time that he can choose out of his assigned roles. For KISS it is sometimes sufficient to avoid this by creating multiple accounts for the few users with multiple roles. Otherwise at least avoid switching roles at run-time in clients as this may cause problems with related states. Simply restart the client with the new role as parameter in case the user wants to switch his role.
+| Access Control | Any permission, group, role, etc., which declares a control for access management.
+|=======================
+
+== Suggestions on the access model
+For the access model we give the following suggestions:
+
+* Each Access Control (permission, group, role, ...) is uniquely identified by a human readable string.
+* We create a unique permission for each use-case.
+* We define groups that combine permissions to typical and useful sets for the users.
+* We define roles as specific groups as required by our business demands.
+* We allow to associate users with a list of Access Controls.
+* For authorization of an implemented use case we determine the required permission. Furthermore, we determine the current user and verify that the required permission is contained in the tree spanned by all his associated Access Controls. If the user does not have the permission we throw a security exception and thus abort the operation and transaction.
+* We avoid negative permissions, that is a user has no permission by default and only those granted to him explicitly give him additional permission for specific things. Permissions granted can not be reduced by other permissions.
+* Technically we consider permissions as a secret of the application. Administrators shall not fiddle with individual permissions but grant them via groups. So the access management provides a list of strings identifying the Access Controls of a user. The individual application itself contains these Access Controls in a structured way, whereas each group forms a permission tree.
+
```

Comment:
Do not use the pattern that no configured permissions mean the user has all permissions